File: Nudge-Neurodivergent-task-tracker/EnergyTracker.java
import java.time.LocalDate;
import java.util.HashMap;
import java.util.Map;
import java.util.List;
import java.util.ArrayList;

/**
 * Tracks daily energy levels and mood to help with task management
 */
public class EnergyTracker {
    public enum EnergyLevel {
        VERY_LOW(1, "😴 Very Low - Rest day vibes"),
        LOW(2, "😌 Low - Gentle tasks only"),
        MODERATE(3, "🙂 Moderate - Steady pace"),
        GOOD(4, "😊 Good - Ready for action"),
        HIGH(5, "⚡ High - Let's tackle everything!");
        
        private final int value;
        private final String description;
        
        EnergyLevel(int value, String description) {
            this.value = value;
            this.description = description;
        }
        
        public int getValue() { return value; }
        public String getDescription() { return description; }
    }
    
    public enum MoodState {
        // Challenging states common in neurodivergent experiences
        OVERWHELMED("😵 Overwhelmed - Need support and simplification"),
        ANXIOUS("😰 Anxious - Seeking safety and predictability"),
        SENSORY_OVERLOAD("🌪️ Sensory overload - Need calm environment"),
        REJECTION_SENSITIVE("💔 RSD activated - Feeling vulnerable"),
        HYPERFOCUS_CRASH("🪫 Post-hyperfocus crash - Energy depleted"),
        EXECUTIVE_DYSFUNCTION("🧠 Executive dysfunction - Can't get started"),
        MASKING_FATIGUE("🎭 Masking fatigue - Socially drained"),
        
        // Neutral and transitional states
        NEUTRAL("😐 Neutral - Regular baseline day"),
        PROCESSING("🤔 Processing - Need time to think"),
        CAUTIOUS("🐌 Cautious - Taking things slow"),
        
        // Positive neurodivergent states
        CONTENT("😌 Content - Comfortable and steady"),
        SPECIAL_INTEREST_MODE("✨ Special interest engaged - Deep focus"),
        HYPERFOCUS_FLOW("🎯 Hyperfocus flow - In the zone"),
        STIMMING_HAPPY("🌈 Stimming happy - Self-regulation working"),
        MOTIVATED("🔥 Motivated - Ready to channel energy");
        
        private final String description;
        
        MoodState(String description) {
            this.description = description;
        }
        
        public String getDescription() { return description; }
        
        public boolean isChallengingState() {
            return this.ordinal() <= 6; // First 7 are challenging states
        }
        
        public boolean isPositiveState() {
            return this.ordinal() >= 10; // Last 5 are positive states
        }
    }
    
    private Map<LocalDate, EnergyLevel> energyHistory;
    private Map<LocalDate, MoodState> moodHistory;
    private Map<LocalDate, String> reflectionNotes;
    private EnergyLevel currentEnergy;
    private MoodState currentMood;
    
    public EnergyTracker() {
        this.energyHistory = new HashMap<>();
        this.moodHistory = new HashMap<>();
        this.reflectionNotes = new HashMap<>();
        this.currentEnergy = EnergyLevel.MODERATE;
        this.currentMood = MoodState.NEUTRAL;
    }
    
    public void logDailyCheckin(EnergyLevel energy, MoodState mood, String reflection) {
        LocalDate today = LocalDate.now();
        energyHistory.put(today, energy);
        moodHistory.put(today, mood);
        
        if (reflection != null && !reflection.trim().isEmpty()) {
            reflectionNotes.put(today, reflection.trim());
        }
        
        this.currentEnergy = energy;
        this.currentMood = mood;
    }
    
    public List<String> getTaskRecommendations() {
        List<String> recommendations = new ArrayList<>();
        
        // Energy-based recommendations
        switch (currentEnergy) {
            case VERY_LOW:
                recommendations.add("🛁 Focus on basic self-care tasks only");
                recommendations.add("🍵 Keep task load minimal - maybe just 1-2 things");
                recommendations.add("🌸 Consider rescheduling non-urgent items");
                recommendations.add("💤 Rest is productive too");
                break;
            case LOW:
                recommendations.add("📝 Tackle simple, low-energy tasks");
                recommendations.add("⏰ Break larger tasks into micro-steps");
                recommendations.add("🎵 Add mood-boosting activities between tasks");
                recommendations.add("🐌 Work at your own gentle pace");
                break;
            case MODERATE:
                recommendations.add("📋 Good day for routine maintenance tasks");
                recommendations.add("⚖️ Balance work and personal items");
                recommendations.add("🎯 Aim for steady, sustainable progress");
                break;
            case GOOD:
                recommendations.add("💪 Great time for challenging or avoided tasks");
                recommendations.add("📞 Handle communication tasks you've been putting off");
                recommendations.add("🧹 Tackle some of those backlogged chores");
                break;
            case HIGH:
                recommendations.add("🚀 Perfect for big projects and hyperfocus sessions");
                recommendations.add("📈 Consider batch processing similar tasks");
                recommendations.add("⚠️ Set timers to avoid burnout from overcommitting");
                break;
        }
        
        // Comprehensive mood-based recommendations
        switch (currentMood) {
            case OVERWHELMED:
                recommendations.add("🫂 Priority: self-soothing and grounding activities");
                recommendations.add("📱 Reach out to someone supportive if possible");
                recommendations.add("🌱 Only do what feels immediately manageable");
                recommendations.add("📋 Consider using a brain dump to externalize thoughts");
                break;
            case ANXIOUS:
                recommendations.add("🧘 Start with calming, predictable activities");
                recommendations.add("📝 Write down worries to clear mental space");
                recommendations.add("🎧 Use familiar background sounds or music");
                recommendations.add("✅ Focus on small, completable tasks for confidence");
                break;
            case SENSORY_OVERLOAD:
                recommendations.add("🌿 Seek quiet, low-stimulation environment");
                recommendations.add("🎧 Use noise-canceling headphones if available");
                recommendations.add("💡 Dim bright lights, reduce visual chaos");
                recommendations.add("🧸 Engage in soothing sensory activities");
                break;
            case REJECTION_SENSITIVE:
                recommendations.add("💝 Avoid tasks requiring vulnerable communication today");
                recommendations.add("🛡️ Focus on solo work where you feel competent");
                recommendations.add("🤗 Practice self-compassion and gentle self-talk");
                recommendations.add("📖 Engage with supportive content or communities");
                break;
            case HYPERFOCUS_CRASH:
                recommendations.add("🔋 Prioritize rest and gentle recharging");
                recommendations.add("🍎 Focus on basic needs: food, water, movement");
                recommendations.add("⏰ Avoid starting new intensive projects");
                recommendations.add("🌸 Be patient with your recovery process");
                break;
            case EXECUTIVE_DYSFUNCTION:
                recommendations.add("🧩 Use external structure: timers, lists, body doubling");
                recommendations.add("⚡ Start with the tiniest possible first step");
                recommendations.add("🔄 Use task-switching if stuck on one thing");
                recommendations.add("📱 Voice memos instead of writing if needed");
                break;
            case MASKING_FATIGUE:
                recommendations.add("🎭 Minimize social demands and interactions");
                recommendations.add("🏠 Create space for authentic self-expression");
                recommendations.add("🌊 Allow natural stims and self-regulation");
                recommendations.add("🔇 Reduce performance pressure on yourself");
                break;
            case SPECIAL_INTEREST_MODE:
                recommendations.add("✨ Channel this focus into relevant tasks if possible");
                recommendations.add("⏰ Set boundaries to prevent complete time loss");
                recommendations.add("🎯 Use this energy for learning or skill-building");
                recommendations.add("💫 Enjoy this state while being mindful of balance");
                break;
            case HYPERFOCUS_FLOW:
                recommendations.add("🎯 Ride this wave but set gentle boundaries");
                recommendations.add("⏰ Use timers for breaks and basic needs");
                recommendations.add("🚀 Tackle your most important or challenging tasks");
                recommendations.add("💧 Remember: hydration, food, and bathroom breaks");
                break;
            case STIMMING_HAPPY:
                recommendations.add("🌈 Maintain this regulation with movement breaks");
                recommendations.add("🎵 Incorporate rhythmic or repetitive tasks");
                recommendations.add("✨ Use this calm alertness for focus work");
                recommendations.add("🔄 Honor your body's need for movement");
                break;
            case PROCESSING:
                recommendations.add("🤔 Give yourself space for slow, thoughtful work");
                recommendations.add("📝 Use written processing over verbal when possible");
                recommendations.add("⏳ Avoid rushed decisions or time pressure");
                recommendations.add("🧠 Trust your need for processing time");
                break;
            case CAUTIOUS:
                recommendations.add("🐌 Start with familiar, low-risk tasks");
                recommendations.add("📋 Make detailed plans to increase confidence");
                recommendations.add("🛡️ Build in extra time and backup plans");
                recommendations.add("✅ Celebrate small wins to build momentum");
                break;
        }
        
        return recommendations;
    }
    
    public String getEnergyTrend() {
        if (energyHistory.size() < 3) {
            return "📊 Not enough data yet - keep tracking!";
        }
        
        List<LocalDate> recentDates = energyHistory.keySet().stream()
            .sorted((d1, d2) -> d2.compareTo(d1))
            .limit(7)
            .toList();
        
        double avgEnergy = recentDates.stream()
            .mapToInt(date -> energyHistory.get(date).getValue())
            .average()
            .orElse(3.0);
        
        if (avgEnergy >= 4.0) {
            return "📈 Your energy has been great lately! Keep up what you're doing!";
        } else if (avgEnergy >= 3.0) {
            return "⚖️ Your energy has been steady. Good balance!";
        } else {
            return "💙 Your energy has been lower lately. Be extra gentle with yourself.";
        }
    }
    
    public void displayDashboard() {
        System.out.println("\n" + "=".repeat(50));
        System.out.println("🌈 ENERGY & MOOD DASHBOARD");
        System.out.println("=".repeat(50));
        
        System.out.println("Current Status:");
        System.out.println("⚡ Energy: " + currentEnergy.getDescription());
        System.out.println("💭 Mood: " + currentMood.getDescription());
        
        System.out.println("\n📊 " + getEnergyTrend());
        
        System.out.println("\n🎯 Today's Recommendations:");
        List<String> recommendations = getTaskRecommendations();
        for (String rec : recommendations) {
            System.out.println("   " + rec);
        }
        
        // Show recent reflection if available
        String todayReflection = reflectionNotes.get(LocalDate.now());
        if (todayReflection != null) {
            System.out.println("\n💭 Today's Reflection:");
            System.out.println("   \"" + todayReflection + "\"");
        }
        
        System.out.println("=".repeat(50));
    }
    
    public boolean shouldReduceTaskLoad() {
        return currentEnergy.getValue() <= 2 || currentMood == MoodState.OVERWHELMED;
    }
    
    public int getRecommendedTaskLimit() {
        switch (currentEnergy) {
            case VERY_LOW: return 1;
            case LOW: return 3;
            case MODERATE: return 5;
            case GOOD: return 7;
            case HIGH: return 10;
            default: return 5;
        }
    }
    
    // Getters
    public EnergyLevel getCurrentEnergy() { return currentEnergy; }
    public MoodState getCurrentMood() { return currentMood; }
    public Map<LocalDate, EnergyLevel> getEnergyHistory() { return new HashMap<>(energyHistory); }
    public Map<LocalDate, MoodState> getMoodHistory() { return new HashMap<>(moodHistory); }
}

File: Nudge-Neurodivergent-task-tracker/energy_tracker.py
"""
Energy and mood tracking system for neurodivergent users
"""

from datetime import date
from enum import Enum
from typing import Dict, List, Optional
from dataclasses import dataclass


class EnergyLevel(Enum):
    """Energy levels with spoon theory reference"""
    VERY_LOW = (1, "😴 Very Low - Rest day vibes")
    LOW = (2, "😌 Low - Gentle tasks only") 
    MODERATE = (3, "🙂 Moderate - Steady pace")
    GOOD = (4, "😊 Good - Ready for action")
    HIGH = (5, "⚡ High - Let's tackle everything!")
    
    def __init__(self, level_value: int, description: str):
        self.level_value = level_value
        self.description = description


class MoodState(Enum):
    """Comprehensive neurodivergent mood states"""
    # Challenging states common in neurodivergent experiences
    OVERWHELMED = ("😵 Overwhelmed - Need support and simplification", True, False)
    ANXIOUS = ("😰 Anxious - Seeking safety and predictability", True, False)
    SENSORY_OVERLOAD = ("🌪️ Sensory overload - Need calm environment", True, False)
    REJECTION_SENSITIVE = ("💔 RSD activated - Feeling vulnerable", True, False)
    HYPERFOCUS_CRASH = ("🪫 Post-hyperfocus crash - Energy depleted", True, False)
    EXECUTIVE_DYSFUNCTION = ("🧠 Executive dysfunction - Can't get started", True, False)
    MASKING_FATIGUE = ("🎭 Masking fatigue - Socially drained", True, False)
    
    # Neutral and transitional states
    NEUTRAL = ("😐 Neutral - Regular baseline day", False, False)
    PROCESSING = ("🤔 Processing - Need time to think", False, False)
    CAUTIOUS = ("🐌 Cautious - Taking things slow", False, False)
    
    # Positive neurodivergent states
    CONTENT = ("😌 Content - Comfortable and steady", False, True)
    SPECIAL_INTEREST_MODE = ("✨ Special interest engaged - Deep focus", False, True)
    HYPERFOCUS_FLOW = ("🎯 Hyperfocus flow - In the zone", False, True)
    STIMMING_HAPPY = ("🌈 Stimming happy - Self-regulation working", False, True)
    MOTIVATED = ("🔥 Motivated - Ready to channel energy", False, True)
    
    def __init__(self, description: str, is_challenging: bool, is_positive: bool):
        self.description = description
        self._is_challenging = is_challenging
        self._is_positive = is_positive
    
    def is_challenging_state(self) -> bool:
        return self._is_challenging
    
    def is_positive_state(self) -> bool:
        return self._is_positive


@dataclass
class DailyCheckin:
    """Daily check-in data"""
    date: date
    energy: EnergyLevel
    mood: MoodState
    reflection: Optional[str] = None


class EnergyTracker:
    """Tracks daily energy levels and mood to help with task management"""
    
    def __init__(self):
        self.energy_history: Dict[date, EnergyLevel] = {}
        self.mood_history: Dict[date, MoodState] = {}
        self.reflection_notes: Dict[date, str] = {}
        self.current_energy = EnergyLevel.MODERATE
        self.current_mood = MoodState.NEUTRAL
    
    def log_daily_checkin(self, energy: EnergyLevel, mood: MoodState, reflection: Optional[str] = None):
        """Log daily check-in data"""
        today = date.today()
        self.energy_history[today] = energy
        self.mood_history[today] = mood
        
        if reflection and reflection.strip():
            self.reflection_notes[today] = reflection.strip()
        
        self.current_energy = energy
        self.current_mood = mood
    
    def get_task_recommendations(self) -> List[str]:
        """Get personalized task recommendations based on current state"""
        recommendations = []
        
        # Energy-based recommendations
        if self.current_energy.level_value == 1:  # VERY_LOW
            recommendations.extend([
                "🛁 Focus on basic self-care tasks only",
                "🍵 Keep task load minimal - maybe just 1-2 things",
                "🌸 Consider rescheduling non-urgent items",
                "💤 Rest is productive too"
            ])
        elif self.current_energy.level_value == 2:  # LOW
            recommendations.extend([
                "📝 Tackle simple, low-energy tasks",
                "⏰ Break larger tasks into micro-steps",
                "🎵 Add mood-boosting activities between tasks",
                "🐌 Work at your own gentle pace"
            ])
        elif self.current_energy.level_value == 3:  # MODERATE
            recommendations.extend([
                "📋 Good day for routine maintenance tasks",
                "⚖️ Balance work and personal items",
                "🎯 Aim for steady, sustainable progress"
            ])
        elif self.current_energy.level_value == 4:  # GOOD
            recommendations.extend([
                "💪 Great time for challenging or avoided tasks",
                "📞 Handle communication tasks you've been putting off",
                "🧹 Tackle some of those backlogged chores"
            ])
        elif self.current_energy.level_value == 5:  # HIGH
            recommendations.extend([
                "🚀 Perfect for big projects and hyperfocus sessions",
                "📈 Consider batch processing similar tasks",
                "⚠️ Set timers to avoid burnout from overcommitting"
            ])
        
        # Comprehensive mood-based recommendations
        mood_recommendations = {
            MoodState.OVERWHELMED: [
                "🫂 Priority: self-soothing and grounding activities",
                "📱 Reach out to someone supportive if possible",
                "🌱 Only do what feels immediately manageable",
                "📋 Consider using a brain dump to externalize thoughts"
            ],
            MoodState.ANXIOUS: [
                "🧘 Start with calming, predictable activities",
                "📝 Write down worries to clear mental space",
                "🎧 Use familiar background sounds or music",
                "✅ Focus on small, completable tasks for confidence"
            ],
            MoodState.SENSORY_OVERLOAD: [
                "🌿 Seek quiet, low-stimulation environment",
                "🎧 Use noise-canceling headphones if available",
                "💡 Dim bright lights, reduce visual chaos",
                "🧸 Engage in soothing sensory activities"
            ],
            MoodState.REJECTION_SENSITIVE: [
                "💝 Avoid tasks requiring vulnerable communication today",
                "🛡️ Focus on solo work where you feel competent",
                "🤗 Practice self-compassion and gentle self-talk",
                "📖 Engage with supportive content or communities"
            ],
            MoodState.HYPERFOCUS_CRASH: [
                "🔋 Prioritize rest and gentle recharging",
                "🍎 Focus on basic needs: food, water, movement",
                "⏰ Avoid starting new intensive projects",
                "🌸 Be patient with your recovery process"
            ],
            MoodState.EXECUTIVE_DYSFUNCTION: [
                "🧩 Use external structure: timers, lists, body doubling",
                "⚡ Start with the tiniest possible first step",
                "🔄 Use task-switching if stuck on one thing",
                "📱 Voice memos instead of writing if needed"
            ],
            MoodState.MASKING_FATIGUE: [
                "🎭 Minimize social demands and interactions",
                "🏠 Create space for authentic self-expression",
                "🌊 Allow natural stims and self-regulation",
                "🔇 Reduce performance pressure on yourself"
            ],
            MoodState.SPECIAL_INTEREST_MODE: [
                "✨ Channel this focus into relevant tasks if possible",
                "⏰ Set boundaries to prevent complete time loss",
                "🎯 Use this energy for learning or skill-building",
                "💫 Enjoy this state while being mindful of balance"
            ],
            MoodState.HYPERFOCUS_FLOW: [
                "🎯 Ride this wave but set gentle boundaries",
                "⏰ Use timers for breaks and basic needs",
                "🚀 Tackle your most important or challenging tasks",
                "💧 Remember: hydration, food, and bathroom breaks"
            ],
            MoodState.STIMMING_HAPPY: [
                "🌈 Maintain this regulation with movement breaks",
                "🎵 Incorporate rhythmic or repetitive tasks",
                "✨ Use this calm alertness for focus work",
                "🔄 Honor your body's need for movement"
            ],
            MoodState.PROCESSING: [
                "🤔 Give yourself space for slow, thoughtful work",
                "📝 Use written processing over verbal when possible",
                "⏳ Avoid rushed decisions or time pressure",
                "🧠 Trust your need for processing time"
            ],
            MoodState.CAUTIOUS: [
                "🐌 Start with familiar, low-risk tasks",
                "📋 Make detailed plans to increase confidence",
                "🛡️ Build in extra time and backup plans",
                "✅ Celebrate small wins to build momentum"
            ]
        }
        
        if self.current_mood in mood_recommendations:
            recommendations.extend(mood_recommendations[self.current_mood])
        
        return recommendations
    
    def get_energy_trend(self) -> str:
        """Get energy trend analysis"""
        if len(self.energy_history) < 3:
            return "📊 Not enough data yet - keep tracking!"
        
        recent_dates = sorted(self.energy_history.keys(), reverse=True)[:7]
        avg_energy = sum(self.energy_history[d].level_value for d in recent_dates) / len(recent_dates)
        
        if avg_energy >= 4.0:
            return "📈 Your energy has been great lately! Keep up what you're doing!"
        elif avg_energy >= 3.0:
            return "⚖️ Your energy has been steady. Good balance!"
        else:
            return "💙 Your energy has been lower lately. Be extra gentle with yourself."
    
    def display_dashboard(self):
        """Display energy and mood dashboard"""
        print("\n" + "=" * 50)
        print("🌈 ENERGY & MOOD DASHBOARD")
        print("=" * 50)
        
        print("Current Status:")
        print(f"⚡ Energy: {self.current_energy.description}")
        print(f"💭 Mood: {self.current_mood.description}")
        
        print(f"\n📊 {self.get_energy_trend()}")
        
        print("\n🎯 Today's Recommendations:")
        recommendations = self.get_task_recommendations()
        for rec in recommendations:
            print(f"   {rec}")
        
        # Show recent reflection if available
        today_reflection = self.reflection_notes.get(date.today())
        if today_reflection:
            print("\n💭 Today's Reflection:")
            print(f'   "{today_reflection}"')
        
        print("=" * 50)
    
    def should_reduce_task_load(self) -> bool:
        """Check if task load should be reduced"""
        return (self.current_energy.level_value <= 2 or 
                self.current_mood == MoodState.OVERWHELMED)
    
    def get_recommended_task_limit(self) -> int:
        """Get recommended task limit based on current energy"""
        energy_limits = {
            1: 1,  # VERY_LOW
            2: 3,  # LOW
            3: 5,  # MODERATE
            4: 7,  # GOOD
            5: 10  # HIGH
        }
        return energy_limits.get(self.current_energy.level_value, 5)
    
    # Getters
    def get_current_energy(self) -> EnergyLevel:
        return self.current_energy
    
    def get_current_mood(self) -> MoodState:
        return self.current_mood
    
    def get_energy_history(self) -> Dict[date, EnergyLevel]:
        return self.energy_history.copy()
    
    def get_mood_history(self) -> Dict[date, MoodState]:
        return self.mood_history.copy()